src/vm/builtins/builtin_runner/bitwise.zig

```diff
@@ -53,14 +93,516 @@ pub const BitwiseBuiltinRunner = struct {
+
+        if (index < self.n_input_cells) {
+            return null;
+        }
+
+        // calculate offset
+        const x_offset = address.subUint(index) catch {
+            return BitwiseError.InvalidBitwiseIndex;
+        };
+        const y_offset = try x_offset.addUint(1);
+
+        const x = try self.getIntWithinBits(x_offset, memory);
+        const y = try self.getIntWithinBits(y_offset, memory);
+
+        const res = switch (index) {
+            2 => x & y, // and
+            3 => x ^ y, // xor
+            4 => x | y, // or
+            else => return BitwiseError.InvalidBitwiseIndex,
+        };
+
+        return .{ .felt = Felt252.fromInteger(res) };
+    }
+
+    /// Retrieves memory segment addresses as a tuple.
+    ///
+    /// Returns a tuple containing the `base` and `stop_ptr` addresses associated
+    /// with the Range Check runner's memory segments. The `stop_ptr` may be `null`.
+    ///
+    /// # Returns
+    /// A tuple of `usize` and `?usize` addresses.
+    pub fn getMemorySegmentAddresses(self: *Self) Tuple(&.{
+        usize,
+        ?usize,
+    }) {
+        return .{
+            self.base,
+            self.stop_ptr,
+        };
+    }
+
+    /// Get the number of used cells associated with this Bitwise runner.
+    ///
+    /// # Parameters
+    ///
+    /// - `segments`: A pointer to a `MemorySegmentManager` for segment size information.
+    ///
+    /// # Returns
+    ///
+    /// The number of used cells as a `u32`, or `MemoryError.MissingSegmentUsedSizes` if
+    /// the size is not available.
+    pub fn getUsedCells(self: *const Self, segments: *MemorySegmentManager) !u32 {
+        return segments.getSegmentUsedSize(
+            @intCast(self.base),
+        ) orelse MemoryError.MissingSegmentUsedSizes;
+    }
+
+    /// Calculates the number of used diluted check units for Keccak hashing.
+    ///
+    /// This function determines the number of used diluted check units based on the
+    /// provided `diluted_n_bits`. It takes into account the allocated virtual columns
+    /// and embedded real cells, providing a count of used check units.
+    ///
+    /// # Parameters
+    /// - `diluted_n_bits`: The number of bits for the diluted check.
+    ///
+    /// # Returns
+    /// The number of used diluted check units as a `usize`.
+    pub fn getUsedDilutedCheckUnits(self: Self, allocator: Allocator, diluted_spacing: u32, diluted_n_bits: u32) !usize {
+        const total_n_bits = self.bitwise_builtin.total_n_bits;
+
+        var partition = std.ArrayList(usize).init(allocator);
+        defer partition.deinit();
+
+        var i: usize = 0;
+
+        while (i < total_n_bits) : (i += diluted_spacing * diluted_n_bits) {
+            for (0..diluted_spacing) |j| {
+                if (i + j < total_n_bits) {
+                    try partition.append(i + j);
+                }
+            }
+        }
+
+        const partition_length = @as(usize, partition.items.len);
```

Comment:
I think that `len` field is already a `usize` natively no? hence maybe no need for `@as(usize,...)` coercion no?